### ISSUE TITLE
EdgeHub's connection status reporeted property becomes empty when device ID contains dot(.)

### ISSUE THREAD

**Issue by yfakariya** [Azure/iotedge#669]
Edge Hub's reported property `clients` does not include any informatin when the device ID including dot (`.`). And that property is shown in Azure portal as "CONNECTED CLIENT COUNT" column of "Modules" tab and "Iot Edge hub connections" tab's modules table in "Edge devices" pane, so they do not show actual values.

This error occured by design flaw which has mismatch amang device/module ID spec, device/module twins' keys spec, and Edge Hub's reported property spec. That is, IoT Hub accepts dot (`.`) and dollar (`$`) for device IDs and module IDs, but module twin's key does not accept such characters. In addition, Edge Hub sends modules' connection statuses with their identifier via `IIdentity.Id` property, and their format are '{deviceID}/{moduleID}' which may include dot and dollar. 

## Expected Behavior

* Module connections reported when device ID contains dot (`.`) and/or dollar (`$`).

## Current Behavior

* Failed to report module connection status for devices which have device ID including dot (`.`). A dollar (`$`) may cause same error potentially.
    * Module ID actually do not have dot/dollar issue because they must be written as key of subproperty of Edge Agent's desired property.
* Edge Hub out warning log to fail to report modules's connection state.

## Steps to Reproduce

1. Create device identity which contains dot, for example "contoso.device1"
2. Add some module such as simulated sensor
3. Check Edge Hub's log via `iotedge logs edgeHub`
4. Check module twin of `$edgeHub` module and connection status via Azure Portal.

### Device (Host) Operating System

Ubuntu server 18.04

### Architecture

amd64

### Container Operating System

Linux

### Runtime Versions

1.0.5

#### iotedged

1.0.5

#### Edge Agent

1.0 (module's log says that the real module version is 1.0.5)

#### Edge Hub

1.0 (module's log says that the real module version is 1.0.5)

## Logs

Here is Edge Hub's warning log:

```
2018-12-14 02:32:14.946 +00:00 [WRN] - Failed to update reported  properties for contoso.device1/$edgeHub with error Property name contoso.device1/module1 contains invalid character '.'
2018-12-14 02:32:14.948 +00:00 [WRN] - Error handling device connected event for device contoso.device1/module1
System.InvalidOperationException: Property name contoso.device1/module1 contains invalid character '.'
   at Microsoft.Azure.Devices.Edge.Hub.Core.TwinManager.ValidatePropertyNameAndLength(String name) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/TwinManager.cs:line 500
   at Microsoft.Azure.Devices.Edge.Hub.Core.TwinManager.ValidateTwinProperties(JToken properties, Int32 currentDepth) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/TwinManager.cs:line 560
   at Microsoft.Azure.Devices.Edge.Hub.Core.TwinManager.ValidateTwinProperties(JToken properties, Int32 currentDepth) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/TwinManager.cs:line 560
   at Microsoft.Azure.Devices.Edge.Hub.Core.TwinManager.UpdateReportedPropertiesWithStoreSupportAsync(String id, IMessage reportedProperties) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/TwinManager.cs:line 457
   at Microsoft.Azure.Devices.Edge.Hub.Core.TwinManager.UpdateReportedPropertiesAsync(String id, IMessage reportedProperties) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/TwinManager.cs:line 379
   at Microsoft.Azure.Devices.Edge.Hub.Core.EdgeHubConnection.DeviceConnected(Object sender, IIdentity device) in /home/vsts/work/1/s/edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Core/EdgeHubConnection.cs:line 277
```

## Additional Information

* Device ID and module ID spec: https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-identity-registry#device-identity-properties (08/29/2018 version)
    * Change of device ID spec is breaking change. So, it should accept dot and dollar forever to keep backward compatibility of data.
    * Module ID spec should be wrong because it must be placed as property key in Edge Agent desired property, so it should not allow dot and dollar as spec. I am afraied that users who use module identity without IoT Edge may be affected when the module ID spec will be changed.
* I am applying IoT Edge to customer who already uses IoT Hub devices and their naming rule for device IDs requires dot to keep relation for their internal database. This issue looks breaking change of device identity.
* In my opinion, it is better to relax device/module twins key charactors.

**Comment by varunpuranik** [comment]
@yfakariya - Thanks for the issue. You have correctly identified what the underlying issue here is - the spec for deviceIds/moduleIds allows characters that the twin does not. This causes issues in reporting the status for the edgeHub/edgeAgent. 
One way to fix this would be to encode those names in the twin. I will mark it as a bug and we can try to fix it soon.

**Comment by ancaantochi** [comment]
Issue was fixed by #682 and will be available with the next release

**Comment by yfakariya** [comment]
Thank you!

**Comment by myagley** [comment]
This is now released in version 1.0.6 of the Edge Hub. You can upgrade by changing the Edge Hub image to `mcr.microsoft.com/azureiotedge-hub:1.0.6`
